### ISSUE TITLE
House CRM API

### ISSUE THREAD

**Issue by estatedocflow** [estatedocflow/estatedocflow.api#7]
**Description:**

This API will provide essential functionalities for managing house data, including property details, attachments, photos, and transaction status.

**API Endpoints Definition:**

**1. Get Houses**
_Endpoint:_ GET /api/houses
_Description:_ Retrieve a list of all houses.
_Response:_ Array of **HouseDto** objects representing each house.

**2. Get House by ID**
_Endpoint:_ GET /api/houses/{id}
_Description:_ Retrieve details of a specific house by its ID.
_Parameters:_ {id}: Unique identifier of the house.
_Response:_ **HouseDto** object representing the house with the specified ID.

**3. Create House**
_Endpoint:_ POST /api/houses
_Description:_ Create a new house entry.
_Request Body:_ **CreateHouseDto** object containing house details.
_Response:_ **HouseDto** object representing the newly created house.

**4. Update House**
_Endpoint:_ PUT /api/houses/{id}
_Description:_ Update all fields of a house identified by its ID.
_Parameters:_ {id}: Unique identifier of the house to update.
_Request Body:_ **UpdateHouseDto** object containing updated house details.
_Response:_ **HouseDto** object representing the updated house.

**5. Partial Update of House (PATCH)**
_Endpoint:_ PATCH /api/houses/{id}
_Description:_ Partially update specific fields of a house identified by its ID.
_Parameters:_ {id}: Unique identifier of the house to update.
_Request Body:_ JSON object containing fields to update (**PatchHouseDto**).
_Response:_ **HouseDto** object representing the updated house.

**6. Delete House**
_Endpoint:_ DELETE /api/houses/{id}
_Description:_ Delete a house by its ID.
_Parameters:_ {id}: Unique identifier of the house to delete.
_Response:_ HTTP status indicating success or failure of the deletion operation.

**7. Patch House Photo**
_Endpoint:_ PATCH /api/houses/{id}/photo
_Description:_ Update or add photos for a specific house.
_Parameters:_ {id}: Unique identifier of the house to update.
_Request Body:_ Array of photo URLs (**PatchPhotoDto**).
_Response_: **HouseDto** object representing the updated house with the new photos.

**8. Patch House Document**
_Endpoint:_ PATCH /api/houses/{id}/document
_Description:_ Update or add documents for a specific house.
_Parameters:_ {id}: Unique identifier of the house to update.
_Request Body:_ Array of document details (**PatchDocumentDto**).
_Response:_ **HouseDto** object representing the updated house with the new documents.

**9. Patch House State**
_Endpoint:_ PATCH /api/houses/{id}/state
_Description:_ Update the state of a house identified by its ID.
_Parameters:_ {id}: Unique identifier of the house to update.
_Request Body:_ JSON object containing the new state (PatchHouseStateDto).
_Response:_ **HouseDto** object representing the updated house.

**Data Transfer Objects (DTOs):**

```
public enum HouseState
{
    [Description("Draft")]
    Draft,

    [Description("For Sale")]
    ForSale,

    [Description("Under Negotiation")]
    UnderNegotiation,

    [Description("Sold")]
    Sold
}

```

```
public class HouseDto
{
    public Guid HouseId { get; set; }
    public string Address { get; set; }
    public string City { get; set; }
    public string OwnerFirstName { get; set; }
    public string OwnerLastName { get; set; }
    public string Description { get; set; }
    public HouseState State { get; set; }
    public List<AttachmentDto> Attachments { get; set; }
    public List<PhotoDto> Photos { get; set; }
}

```
```
public class CreateHouseDto
{
    public string Address { get; set; }
    public string City { get; set; }
    public string OwnerFirstName { get; set; }
    public string OwnerLastName { get; set; }
    public string Description { get; set; }
}
```

```
public class UpdateHouseDto
{
    public string Address { get; set; }
    public string City { get; set; }
    public string OwnerFirstName { get; set; }
    public string OwnerLastName { get; set; }
    public string Description { get; set; }
}
```

```
public class PatchPhotoDto
{
    public List<string> PhotoUrls { get; set; }
}
```

```
public class PatchDocumentDto
{
    public List<DocumentDto> Documents { get; set; }
}
```

```
public class DocumentDto
{
    public string DocumentType { get; set; }
    public string DocumentUrl { get; set; }
}
```

```
public class PatchHouseStateDto
{
    public HouseState State { get; set; }
}
```

**Comment by estatedocflow** [comment]
@Riya2993 sign here working hours, please!

You missed also the PR to merge it in the dev branch, please added it.

**Comment by estatedocflow** [comment]
@Riya2993 I saw your changes, please don't change the SDK version, use .NET 8.

**Comment by Riya2993** [comment]
I have added the entity for House and HousePhoto 
public class House
    {
        [Key]
        public Guid Id { get; set; }  
        public string Address { get; set; }
        public string City { get; set; }
        public string OwnerFirstName { get; set; }
        public string OwnerLastName { get; set; }
        public string Description { get; set; }
        public HouseState State { get; set; }
        public virtual IEnumerable<HousePhoto>? HousePhotos { get; set; }
    }
    
    public class HousePhoto
    {
        [Key]
        public Guid Id { get; set; }

        [Required]
        [ForeignKey("House")]
        public Guid HouseId { get; set; }
        public string PhotoUrl { get; set; }
        public virtual House House { get; set; }
    }
    
Also, I cannot see the AttachmentDto, just can see its use in the HouseDto model.


Merged Main into Dev with latest changes

**Comment by estatedocflow** [comment]
@Riya2993 you're right, you can delete the HousePhoto model and create a generic HouseAttachment, with an additional field called DocumentType (it will be an enum, and at the moment we only have Photo - 0, Document - 1)

**Comment by Riya2993** [comment]
DocumentType save as 0 and 1 in HouseAttachment table ?

**Comment by Riya2993** [comment]
public class House
    {
        [Key]
        public Guid Id { get; set; }  
        public string Address { get; set; }
        public string City { get; set; }
        public string OwnerFirstName { get; set; }
        public string OwnerLastName { get; set; }
        public string Description { get; set; }
        public HouseState State { get; set; }
        public virtual IEnumerable<HouseAttachment>? HouseAttachments { get; set; }
    }
    
    public class HouseAttachment
    {
        [Key]
        public Guid Id { get; set; }

        [Required]
        [ForeignKey("House")]
        public Guid HouseId { get; set; }
        public string PhotoUrl { get; set; }
        public DocumentType DocumentType { get; set; }
        public virtual Models.Entities.House.House House { get; set; }
    }
    
   It is correct entity now ?

**Comment by estatedocflow** [comment]
yes is correct, ok 0 and 1 is enought

**Comment by Riya2993** [comment]
7. Patch House Photo
8. Patch House Document

How we need to generate Image and Document Url?
Also from request  in which type of  format Url get?

**Comment by Riya2993** [comment]
5. Partial Update of House (PATCH)
Endpoint: PATCH /api/houses/{id}
Description: Partially update specific fields of a house identified by its ID.
Parameters: {id}: Unique identifier of the house to update.
Request Body: JSON object containing fields to update (PatchHouseDto).
Response: HouseDto object representing the updated house.

PatchHouseDto is not defined and which specific fields is required to update.

**Comment by estatedocflow** [comment]
@Riya2993 Patch House Photo
Patch House Document
How we need to generate Image and Document Url? 
Also from request in which type of format Url get? 

we'll store the document and image in an AWS S3 bucket, so it will be a GUID. WDYT?

Partial Update of House (PATCH)
Endpoint: PATCH /api/houses/{id}
Description: Partially update specific fields of a house identified by its ID.
Parameters: {id}: Unique identifier of the house to update.
Request Body: JSON object containing fields to update (PatchHouseDto).
Response: HouseDto object representing the updated house.
PatchHouseDto is not defined and which specific fields is required to update.

You're right... it is a refuse... ignore it. Thanks

**Comment by Riya2993** [comment]
To store documents in Amazon s3 bucket, we need AWS credentials(like accessKey and secreyKey) and the name of the bucket. So, after storing Image in document we'll get ImageUrl(string).

**Comment by estatedocflow** [comment]
@Riya2993 yeas of course, but we'll do this in another issue, at the moment you only build the structure to do that. ok?

**Comment by Riya2993** [comment]
Total hours for implementation: 4

**Comment by Riya2993** [comment]
Committed, tested and updated the balance hours.
Additional hours - 2.

**Comment by Riya2993** [comment]
Additional hours spent for resolving - 2
